dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyAbstractPeriodType.java

```diff
@@ -0,0 +1,152 @@
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.period;
+
+import java.util.Date;
+import java.util.List;
+
+import org.hisp.dhis.calendar.Calendar;
+import org.hisp.dhis.calendar.DateTimeUnit;
+
+import com.google.common.collect.Lists;
+
+/**
+ * @author Abyot Asalefew Gizaw <abyota@gmail.com>
+ *
+ */
+public abstract class QuarterlyAbstractPeriodType
+    extends CalendarPeriodType
+{
+    public static final int FREQUENCY_ORDER = 91;
+
+    private static final String ISO8601_DURATION = "P3M";
+
+    public static final String SQL_INTERVAL = "3 months";
+
+    protected static final String ISO_CALENDAR_NAME = org.hisp.dhis.calendar.impl.Iso8601Calendar.getInstance().name();
+
+    // -------------------------------------------------------------------------
+    // PeriodType functionality
+    // -------------------------------------------------------------------------
+
+    @Override
+    public int getFrequencyOrder()
+    {
+        return FREQUENCY_ORDER;
+    }
+
+    @Override
+    public String getSqlInterval()
+    {
+        return SQL_INTERVAL;
+    }
+
+    @Override
+    public String getIso8601Duration()
+    {
+        return ISO8601_DURATION;
+    }
+
+    // -------------------------------------------------------------------------
+    // CalendarPeriodType functionality
+    // -------------------------------------------------------------------------
+
+    @Override
+    public DateTimeUnit getDateWithOffset( DateTimeUnit dateTimeUnit, int offset, Calendar calendar )
+    {
+        return calendar.plusMonths( dateTimeUnit, offset * 3 );
+    }
+
+    /**
+     * Generates quarterly Periods for the whole year in which the given
+     * Period's startDate exists.
+     */
+    @Override
+    public List<Period> generatePeriods( DateTimeUnit dateTimeUnit )
+    {
+        org.hisp.dhis.calendar.Calendar cal = getCalendar();
+
+        dateTimeUnit.setMonth( 1 );
+        dateTimeUnit.setDay( 1 );
+
+        int year = dateTimeUnit.getYear();
+        List<Period> periods = Lists.newArrayList();
+
+        while ( year == dateTimeUnit.getYear() )
+        {
+            periods.add( createPeriod( dateTimeUnit, cal ) );
+            dateTimeUnit = cal.plusMonths( dateTimeUnit, 3 );
+        }
+
+        return periods;
+    }
+
+    /**
+     * Generates the last 4 quarters where the last one is the quarter which the
+     * given date is inside.
+     */
+    @Override
+    public List<Period> generateRollingPeriods( Date date )
+    {
+        date = createPeriod( date ).getStartDate();
+
+        return generateRollingPeriods( createLocalDateUnitInstance( date ), getCalendar() );
+    }
+
+    @Override
+    public List<Period> generateRollingPeriods( DateTimeUnit dateTimeUnit, Calendar calendar )
```

Comment:
Would be nice to have, in the Javadoc, documentation for the params and the return object.